File: MuscleHamster/Models/UserProfile.swift
//
//  UserProfile.swift
//  MuscleHamster
//
//  User profile model storing onboarding preferences for workout personalization
//

import Foundation

// MARK: - User Profile

struct UserProfile: Codable, Equatable {
    var age: Int?
    var fitnessLevel: FitnessLevel?
    var fitnessGoals: Set<FitnessGoal>
    var weeklyWorkoutGoal: Int?
    var schedulePreference: SchedulePreference?
    var preferredWorkoutTime: WorkoutTime?
    var fitnessIntent: FitnessIntent?
    var hamsterName: String?

    /// Current onboarding step (0-indexed). Used to resume after abandonment.
    var currentStep: Int

    init(
        age: Int? = nil,
        fitnessLevel: FitnessLevel? = nil,
        fitnessGoals: Set<FitnessGoal> = [],
        weeklyWorkoutGoal: Int? = nil,
        schedulePreference: SchedulePreference? = nil,
        preferredWorkoutTime: WorkoutTime? = nil,
        fitnessIntent: FitnessIntent? = nil,
        hamsterName: String? = nil,
        currentStep: Int = 0
    ) {
        self.age = age
        self.fitnessLevel = fitnessLevel
        self.fitnessGoals = fitnessGoals
        self.weeklyWorkoutGoal = weeklyWorkoutGoal
        self.schedulePreference = schedulePreference
        self.preferredWorkoutTime = preferredWorkoutTime
        self.fitnessIntent = fitnessIntent
        self.hamsterName = hamsterName
        self.currentStep = currentStep
    }

    /// Whether all required fields are filled
    var isComplete: Bool {
        age != nil &&
        fitnessLevel != nil &&
        !fitnessGoals.isEmpty &&
        weeklyWorkoutGoal != nil &&
        schedulePreference != nil &&
        preferredWorkoutTime != nil &&
        fitnessIntent != nil &&
        hamsterName != nil && !hamsterName!.isEmpty
    }

    // MARK: - Hamster Name Validation

    /// Minimum length for hamster name
    static let hamsterNameMinLength = 1

    /// Maximum length for hamster name
    static let hamsterNameMaxLength = 24

    /// Validates the hamster name and returns an error message if invalid
    static func validateHamsterName(_ name: String) -> String? {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)

        if trimmed.isEmpty {
            return "Your hamster needs a name!"
        }

        if trimmed.count > hamsterNameMaxLength {
            return "That name is a bit long. Try \(hamsterNameMaxLength) characters or less."
        }

        // Allow letters, numbers, spaces, and common punctuation
        let allowedCharacters = CharacterSet.alphanumerics
            .union(.whitespaces)
            .union(CharacterSet(charactersIn: "-'"))

        if trimmed.unicodeScalars.contains(where: { !allowedCharacters.contains($0) }) {
            return "Please use only letters, numbers, spaces, or hyphens."
        }

        return nil
    }

    /// Whether the hamster name is valid
    static func isValidHamsterName(_ name: String) -> Bool {
        validateHamsterName(name) == nil
    }
}

// MARK: - Fitness Level

enum FitnessLevel: String, Codable, CaseIterable, Identifiable {
    case beginner
    case intermediate
    case advanced

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .beginner: return "Beginner"
        case .intermediate: return "Intermediate"
        case .advanced: return "Advanced"
        }
    }

    var description: String {
        switch self {
        case .beginner: return "New to fitness or getting back into it"
        case .intermediate: return "Regular workouts for a few months"
        case .advanced: return "Consistent training for a year or more"
        }
    }
}

// MARK: - Fitness Goals

enum FitnessGoal: String, Codable, CaseIterable, Identifiable {
    case cardio
    case muscleGain
    case fatLoss
    case flexibility
    case generalFitness

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .cardio: return "Cardio & Endurance"
        case .muscleGain: return "Build Muscle"
        case .fatLoss: return "Lose Weight"
        case .flexibility: return "Flexibility & Mobility"
        case .generalFitness: return "General Fitness"
        }
    }

    var icon: String {
        switch self {
        case .cardio: return "heart.fill"
        case .muscleGain: return "dumbbell.fill"
        case .fatLoss: return "flame.fill"
        case .flexibility: return "figure.flexibility"
        case .generalFitness: return "figure.run"
        }
    }
}

// MARK: - Schedule Preference

enum SchedulePreference: String, Codable, CaseIterable, Identifiable {
    case fixed
    case flexible

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .fixed: return "Fixed Schedule"
        case .flexible: return "Flexible Schedule"
        }
    }

    var description: String {
        switch self {
        case .fixed: return "Same days each week (e.g., Mon, Wed, Fri)"
        case .flexible: return "Any days that work for you"
        }
    }
}

// MARK: - Workout Time

enum WorkoutTime: String, Codable, CaseIterable, Identifiable {
    case morning
    case afternoon
    case evening
    case noPreference

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .morning: return "Morning"
        case .afternoon: return "Afternoon"
        case .evening: return "Evening"
        case .noPreference: return "No Preference"
        }
    }

    var timeRange: String {
        switch self {
        case .morning: return "Before noon"
        case .afternoon: return "12pm - 5pm"
        case .evening: return "After 5pm"
        case .noPreference: return "Whenever works"
        }
    }

    var icon: String {
        switch self {
        case .morning: return "sun.horizon.fill"
        case .afternoon: return "sun.max.fill"
        case .evening: return "moon.fill"
        case .noPreference: return "clock.fill"
        }
    }
}

// MARK: - Fitness Intent

enum FitnessIntent: String, Codable, CaseIterable, Identifiable {
    case maintenance
    case improvement

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .maintenance: return "Maintain"
        case .improvement: return "Improve"
        }
    }

    var description: String {
        switch self {
        case .maintenance: return "Keep my current fitness level steady"
        case .improvement: return "Push myself to get stronger and fitter"
        }
    }
}

// MARK: - UserDefaults Persistence

extension UserProfile {
    private static let storageKey = "partialOnboardingProfile"

    /// Save partial progress to UserDefaults
    func saveToUserDefaults() {
        if let encoded = try? JSONEncoder().encode(self) {
            UserDefaults.standard.set(encoded, forKey: Self.storageKey)
        }
    }

    /// Load partial progress from UserDefaults
    static func loadFromUserDefaults() -> UserProfile? {
        guard let data = UserDefaults.standard.data(forKey: storageKey),
              let profile = try? JSONDecoder().decode(UserProfile.self, from: data) else {
            return nil
        }
        return profile
    }

    /// Clear saved progress from UserDefaults
    static func clearFromUserDefaults() {
        UserDefaults.standard.removeObject(forKey: storageKey)
    }
}

File: MuscleHamsterExpo/src/models/UserProfile.js
// User Profile Model - Simplified
// Focused on data that actually influences the app experience

// ============================================
// FITNESS LEVEL
// ============================================

export const FitnessLevel = {
  BEGINNER: 'beginner',
  INTERMEDIATE: 'intermediate',
  ADVANCED: 'advanced',
};

export const FitnessLevelInfo = {
  [FitnessLevel.BEGINNER]: {
    displayName: 'Beginner',
    description: 'New to fitness or getting back into it',
    icon: 'leaf',
  },
  [FitnessLevel.INTERMEDIATE]: {
    displayName: 'Intermediate',
    description: 'Work out regularly, ready for more',
    icon: 'fitness',
  },
  [FitnessLevel.ADVANCED]: {
    displayName: 'Advanced',
    description: 'Experienced and pushing limits',
    icon: 'trophy',
  },
};

// ============================================
// FITNESS GOALS (Outcome-focused)
// ============================================

export const FitnessGoal = {
  STAY_ACTIVE: 'stayActive',
  GET_STRONGER: 'getStronger',
  BUILD_CORE: 'buildCore',
  IMPROVE_FLEXIBILITY: 'improveFlexibility',
};

export const FitnessGoalInfo = {
  [FitnessGoal.STAY_ACTIVE]: {
    displayName: 'Stay Active',
    description: 'Keep moving with quick, easy workouts',
    icon: 'flash',
    // Maps to: Quick Sweats, Desk Workouts
    categories: ['quick_sweats', 'desk'],
  },
  [FitnessGoal.GET_STRONGER]: {
    displayName: 'Get Stronger',
    description: 'Build muscle and increase strength',
    icon: 'barbell',
    // Maps to: Upper Body, Lower Body, Gym workouts
    categories: ['upper_body', 'lower_body', 'legs', 'arms', 'back', 'chest', 'shoulders'],
  },
  [FitnessGoal.BUILD_CORE]: {
    displayName: 'Build Core',
    description: 'Strengthen your core and abs',
    icon: 'ellipse',
    // Maps to: Core
    categories: ['core'],
  },
  [FitnessGoal.IMPROVE_FLEXIBILITY]: {
    displayName: 'Improve Flexibility',
    description: 'Stretch more and move better',
    icon: 'body',
    // Maps to: Flexibility/Yoga content
    categories: ['flexibility', 'yoga', 'stretching'],
  },
};

// ============================================
// WORKOUT DAYS
// ============================================

export const WeekDay = {
  MONDAY: 'monday',
  TUESDAY: 'tuesday',
  WEDNESDAY: 'wednesday',
  THURSDAY: 'thursday',
  FRIDAY: 'friday',
  SATURDAY: 'saturday',
  SUNDAY: 'sunday',
};

export const WeekDayInfo = {
  [WeekDay.MONDAY]: { short: 'Mon', letter: 'M' },
  [WeekDay.TUESDAY]: { short: 'Tue', letter: 'T' },
  [WeekDay.WEDNESDAY]: { short: 'Wed', letter: 'W' },
  [WeekDay.THURSDAY]: { short: 'Thu', letter: 'T' },
  [WeekDay.FRIDAY]: { short: 'Fri', letter: 'F' },
  [WeekDay.SATURDAY]: { short: 'Sat', letter: 'S' },
  [WeekDay.SUNDAY]: { short: 'Sun', letter: 'S' },
};

export const WEEK_DAYS_ORDERED = [
  WeekDay.MONDAY,
  WeekDay.TUESDAY,
  WeekDay.WEDNESDAY,
  WeekDay.THURSDAY,
  WeekDay.FRIDAY,
  WeekDay.SATURDAY,
  WeekDay.SUNDAY,
];

// ============================================
// PROFILE STRUCTURE
// ============================================

// Default empty profile
export const createEmptyProfile = () => ({
  // Essential
  hamsterName: null,

  // Fitness preferences (simplified)
  fitnessLevel: null,
  fitnessGoals: [], // Array of FitnessGoal values
  workoutDays: [],  // Array of WeekDay values

  // Profile state
  profileComplete: false,
  profileVersion: 2, // Version 2 = simplified profile

  // Legacy fields (kept for migration)
  // These will be ignored but preserved for existing users
});

// Check if profile needs migration (old format)
export const needsProfileMigration = (profile) => {
  if (!profile) return false;

  // If profile has old fields but no profileVersion or version 1
  const hasOldFields = profile.schedulePreference !== undefined ||
                       profile.preferredWorkoutTime !== undefined ||
                       profile.fitnessIntent !== undefined ||
                       profile.weeklyWorkoutGoal !== undefined;

  const isOldVersion = !profile.profileVersion || profile.profileVersion < 2;

  return hasOldFields && isOldVersion && profile.profileComplete;
};

// Migrate old profile to new format
export const migrateProfile = (oldProfile) => {
  if (!oldProfile) return createEmptyProfile();

  // Map old fitness goals to new ones
  const oldGoals = oldProfile.fitnessGoals || [];
  const newGoals = [];

  // Map old goal types to new
  if (oldGoals.includes('cardio') || oldGoals.includes('generalFitness')) {
    newGoals.push(FitnessGoal.STAY_ACTIVE);
  }
  if (oldGoals.includes('buildMuscle')) {
    newGoals.push(FitnessGoal.GET_STRONGER);
  }
  if (oldGoals.includes('loseFat')) {
    // Map to both stay active and build core
    if (!newGoals.includes(FitnessGoal.STAY_ACTIVE)) {
      newGoals.push(FitnessGoal.STAY_ACTIVE);
    }
    newGoals.push(FitnessGoal.BUILD_CORE);
  }
  if (oldGoals.includes('flexibility')) {
    newGoals.push(FitnessGoal.IMPROVE_FLEXIBILITY);
  }

  // Convert weeklyWorkoutGoal to workout days
  // Default to Mon/Wed/Fri pattern
  let workoutDays = [];
  const weeklyGoal = oldProfile.weeklyWorkoutGoal || 3;

  if (weeklyGoal >= 1) workoutDays.push(WeekDay.MONDAY);
  if (weeklyGoal >= 2) workoutDays.push(WeekDay.WEDNESDAY);
  if (weeklyGoal >= 3) workoutDays.push(WeekDay.FRIDAY);
  if (weeklyGoal >= 4) workoutDays.push(WeekDay.TUESDAY);
  if (weeklyGoal >= 5) workoutDays.push(WeekDay.THURSDAY);
  if (weeklyGoal >= 6) workoutDays.push(WeekDay.SATURDAY);
  if (weeklyGoal >= 7) workoutDays.push(WeekDay.SUNDAY);

  return {
    hamsterName: oldProfile.hamsterName,
    fitnessLevel: oldProfile.fitnessLevel,
    fitnessGoals: newGoals.length > 0 ? newGoals : [FitnessGoal.STAY_ACTIVE],
    workoutDays: workoutDays,
    profileComplete: false, // Mark incomplete so they go through new flow
    profileVersion: 2,
  };
};

// ============================================
// HAMSTER NAME VALIDATION
// ============================================

export const HAMSTER_NAME_MIN_LENGTH = 1;
export const HAMSTER_NAME_MAX_LENGTH = 24;
export const HAMSTER_NAME_PATTERN = /^[a-zA-Z0-9\s\-']+$/;

export const validateHamsterName = (name) => {
  if (!name || name.trim().length === 0) {
    return { valid: false, error: 'Your hamster needs a name!' };
  }
  if (name.length > HAMSTER_NAME_MAX_LENGTH) {
    return { valid: false, error: `Name must be ${HAMSTER_NAME_MAX_LENGTH} characters or less` };
  }
  if (!HAMSTER_NAME_PATTERN.test(name)) {
    return { valid: false, error: 'Only letters, numbers, spaces, hyphens, and apostrophes allowed' };
  }
  return { valid: true, error: null };
};

export const HAMSTER_NAME_SUGGESTIONS = ['Hammy', 'Peanut', 'Whiskers', 'Nugget', 'Biscuit', 'Coco'];

// ============================================
// LEGACY EXPORTS (for backward compatibility)
// These are kept so existing code doesn't break
// ============================================

export const SchedulePreference = {
  FIXED: 'fixed',
  FLEXIBLE: 'flexible',
};

export const SchedulePreferenceInfo = {
  [SchedulePreference.FIXED]: {
    displayName: 'Fixed Days',
    description: 'Same days each week',
    icon: 'calendar',
  },
  [SchedulePreference.FLEXIBLE]: {
    displayName: 'Flexible',
    description: 'Any days work',
    icon: 'shuffle',
  },
};

export const WorkoutTime = {
  MORNING: 'morning',
  AFTERNOON: 'afternoon',
  EVENING: 'evening',
  NO_PREFERENCE: 'noPreference',
};

export const WorkoutTimeInfo = {
  [WorkoutTime.MORNING]: { displayName: 'Morning', description: '6am - 12pm', icon: 'sunny' },
  [WorkoutTime.AFTERNOON]: { displayName: 'Afternoon', description: '12pm - 5pm', icon: 'partly-sunny' },
  [WorkoutTime.EVENING]: { displayName: 'Evening', description: '5pm - 10pm', icon: 'moon' },
  [WorkoutTime.NO_PREFERENCE]: { displayName: 'No Preference', description: 'Whenever works!', icon: 'time' },
};

export const FitnessIntent = {
  MAINTAIN: 'maintain',
  IMPROVE: 'improve',
};

export const FitnessIntentInfo = {
  [FitnessIntent.MAINTAIN]: { displayName: 'Maintain', description: 'Keep current level', icon: 'checkmark-circle' },
  [FitnessIntent.IMPROVE]: { displayName: 'Improve', description: 'Push to get better', icon: 'trending-up' },
};
